katsdpimageutils/primary_beam_correction.py

```diff
@@ -0,0 +1,274 @@
+
+
+def cosine_power_pattern(separation_rad, c_freq):
+    """Compute Power patterns for a given frequency.
+
+    This uses the Cosine-squared power approximation from
+    Mauch et al. (2020).
+
+    Parameters
+    ----------
+    separation_rad : numpy array
+        Radial separation array
+    c_freq : numpy array
+        An array of central frequencies for each frequency plane
+    """
+    rho = separation_rad
+    v_beam_rad = np.deg2rad(89.5 / 60.)
+    h_beam_rad = np.deg2rad(86.2 / 60.)
+    # Take the Geometric mean for the vertical and horizontal cut through the beam.
+    vh_beam_mean = np.sqrt(v_beam_rad * h_beam_rad)
+    flux_density = []
+    for nu in c_freq:
+        # Convert GHz to Hz
+        nu = nu/1.e9
+        theta_b = vh_beam_mean / nu
+        ratio = rho/theta_b
+        num = np.cos(1.189 * np.pi * ratio)
+        den = 1-4 * (1.189 * ratio)**2
+        a_b = (num/den)**2
+        flux_density.append(a_b)
+    return flux_density
+
+
+def beam_pattern(path):
+    """Make beam pattern.
+
+    Parameters
+    ----------
+    path : str
+        FITS file
+    """
+    # Read the fits file
+    data = read_fits(path)
+    # Get the central frequency of the image header given.
+    c_freq = central_freq(path)
+    # Get radial separation between sources and the phase centre
+    phase_center, image_wcs = get_position(path)
+    separation_rad = radial_offset(phase_center, image_wcs)
+    beam_list = cosine_power_pattern(separation_rad, c_freq)
+    return beam_list, data
+
+
+def standard_deviation(data):
+    """Compute the median and the estimate of the standard deviation.
+
+    This is based on the median absolute deviation (MAD).
+    """
+    MAD_TO_SD = 1.4826
+    med = np.nanmedian(data)
+    dev = np.abs(data - med)
+    return med, MAD_TO_SD * np.nanmedian(dev)
+
+
+def inverse_variance(data):
+    """Calculate the inverse variance.
+
+    Reject pixels more than 5 sigma from the median and reject all zeros
+    until either no more pixels are rejected or a maximum of 50 iterations
+    is reached.
+    """
+    data = data[data != 0.0]
+    if len(data) == 0:
+        return 0.0
+    med, sd = standard_deviation(data)
+    for i in range(50):
+        old_sd = sd
+        cut = np.abs(data - med) < 5.0 * sd
+        if np.all(~cut):
+            return 1/(sd)**2
+        data = data[cut]
+        med, sd = standard_deviation(data)
+        if sd == 0.0:
+            return 1/(old_sd)**2
+    return 1/(sd)**2
+
+
+def weighted_average(arr, weights):
+    """Compute weighted average of all the frequency planes."""
+    wt_average = np.average(arr, weights=weights, axis=0)
+    return wt_average
+
+
+def primary_beam_correction(beam_pattern, raw_image, px_cut=0.1):
+    """Correct the effects of primary beam.
+
+    Parameters
+    ----------
+    beam_pattern : numpy array
+        Array of beam pattern
+    raw_image : astropy.wcs.wcs.WCS
+        WCS keywords in the primary HDU
```

Comment:
This still looks like the documentation doesn't match the actual parameter.

Reply:
Resolved.